Loosen SparseMatrixCSC colptr and rowval integer type restriction

Pbellive opened the pull request:
This fixes #31435. Please see that issue for full explanation. @KlausC any objection to this change?

Commit:
Make SparseMatrixCSC colptr and rowval integer type restriction less restrictive. Integer type needs to be wide enough to store matrix with number of non-zeros given, not any sparse matrix of equal size.

ViralBShah commented:
I noted a potential objection in #31435.

KlausC reviewed (commented):
should be: `nnzval < typemax(Ti)` for correct handling of corner case `Ti = Int; nnzval = typemax(Ti)`.

error text `"Number .."` should be lowercase for consistency with other messages.

fredrikekre commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -42,7 +42,7 @@ function sparse_check_Ti(m::Integer, n::Integer, Ti::Type)
         !isbitstype(Ti) || m ≤ typemax(Ti) || throwTi("number of rows", "m", m)
         !isbitstype(Ti) || n ≤ typemax(Ti) || throwTi("number of columns", "n", n)
-        !isbitstype(Ti) || n*m+1 ≤ typemax(Ti) || throwTi("maximal nnz+1", "m*n+1", n*m+1)
+        !isbitstype(Ti) || nnzval+1 ≤ typemax(Ti) || throwTi("Number of nonzeros", "nnz+1", nnzval+1)
```
```suggestion
        !isbitstype(Ti) || nnzval < typemax(Ti) || throwTi("number of nonzeros", "nnz+1", nnzval+1)
```

KlausC commented:
As `nnz(A)` may change during the lifetime of `A`, it would be wise to assert that the value we want to store in `A.colptr[A.n+1]`  (that is the correct new `nnz(A)+1`) fits in a `Ti`.

KlausC commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -23,7 +23,7 @@ struct SparseMatrixCSC{Tv,Ti<:Integer} <: AbstractSparseMatrix{Tv,Ti}
                                     nzval::Vector{Tv}) where {Tv,Ti<:Integer}
 
-        sparse_check_Ti(m, n, Ti)
+        sparse_check_Ti(m, n, length(nzval), Ti)
```
```suggestion
        sparse_check_Ti(m, n, max(length(nzval), length(rowval)), Ti)
```

KlausC commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -594,7 +594,7 @@ function sparse!(I::AbstractVector{Ti}, J::AbstractVector{Ti},
 
     require_one_based_indexing(I, J, V)
-    sparse_check_Ti(m, n, Ti)
+    sparse_check_Ti(m, n, length(V), Ti)
```
```suggestion
    sparse_check_Ti(m, n, length(I), Ti)
```

ViralBShah commented:
Can't `setindex!` potentially grow the size of `nzval` and `rowval`, thus requiring the check to be the way it is right now (but being careful about the overflow)?

mbauman commented:
Yes, but:

* That's not how things behaved prior to #31118
* The change in #31118 was breaking
* We merged #31118 without consideration of the breaking-ness

So these sorts of "it'd be nice to guarantee"s are non-sequitors.  We just need to get things back to the old status quo and then we can discuss how to approach the next steps.

It's either merge this patch or wholesale revert #31118.

KristofferC commented:
The discussion here shows that there are intricacies we don't really have time to deal with now, considering we want to branch for 1.3 about now. I suggest just reverting the breaking PR (https://github.com/JuliaLang/julia/pull/31667) and then the discussion can take as long as it wants to add it back in a non-breaking way.

KlausC commented:
Sorry, but without a replacement for #31118 the software is breaking because of #31024. Propose to merge this PR with the proposed changes.

KristofferC commented:
> the software is breaking because of

What code would reverting the PR break? I see an error being replaced with another one.

mbauman commented:
Note that we're using breaking with a very specific meaning here: it's not the Julia is "broken" (e.g., with a seg fault in some instances) but rather that a particular change caused a regression that broke someone's workflow that previously worked.  That's significantly worse than an existing bug that has been a bug for a long time.

Pbellive commented:
I'm in favour of @KristofferC 's suggestion to revert #31118 and continue discussion of a long term fix here. @ViralBShah and @KlausC 's comments that `setindex!` can grow the internal arrays are of course important points. To address that, it seems like it would be sufficient to just add a type check inside the setindex! methods to make sure that `A.colptr[n+1] < typemax(Ti)` remains true after the new entries are added to the matrix. I haven't tested that but I wouldn't think it would be much of a performance hit.

mauro3 commented:
> Sorry, but without a replacement for #31118 the software is breaking because of #31024

@KlausC: until it is fixed you can monkey-patch it: https://discourse.julialang.org/t/difficulties-with-recent-julia-releases/22969/2)

KlausC commented:
@mbauman 
> Note that we're using breaking with a very specific meaning here: ... a particular change caused a regression that broke someone's workflow that previously worked

Does that mean, we don't introduce data consistency checks, because someone's program worked, although the data, it operated on, were partially corrupted?

In our case, I agree, that the introduced check was too restrictive, but generally ...?

mbauman commented:
It just means that we need to be careful about doing so.  We flag all such changes as [minor changes](https://github.com/JuliaLang/julia/issues?q=label%3A"minor+change"), we ensure they appear in NEWS.md, and such changes have a higher bar to cross to get merged.

In cases where the previous behavior resulted in corrupted or buggy answers, then that's not breaking — it's a bug.  Admittedly sometimes the line between the two can be blurry.

In this case, I still really want to see a resurrected #31118+#31661 that fixes the segfault and protects against overfilling such arrays.  Had we not been right on a branching threshold, we probably would have let things percolate a little longer without a direct revert.

ViralBShah commented:
In general, we always want correctness, but this was a tricky situation as already discussed above. Reverting and redoing is certainly the right thing.

Pbellive commented:
I'm happy to update and expand this PR to make an attempt at addressing the concerns raised in this thread. I should be able to get to it in the next 4-5 days.

KlausC commented:
@Pbellive sorry for taking your job. I hope you can live with this solution.
